src/cata_imgui.cpp

```diff
@@ -0,0 +1,466 @@
+}
+
+void cataimgui::client::end_frame()
+{
+    ImGui::Render();
+    ImGui_ImplSDLRenderer2_RenderDrawData( ImGui::GetDrawData() );
+}
+
+void cataimgui::client::process_input( void *input )
+{
+    ImGui_ImplSDL2_ProcessEvent( static_cast<const SDL_Event *>( input ) );
+}
+
+#endif
+
+void cataimgui::window::draw_colored_text( std::string const &text, const nc_color &color,
+        text_align alignment, float max_width, bool *is_selected, bool *is_focused, bool *is_hovered )
+{
+    nc_color color_cpy = color;
+    draw_colored_text( text, color_cpy, alignment, max_width, is_selected, is_focused, is_hovered );
+}
+
+void cataimgui::window::draw_colored_text( std::string const &text, nc_color &color,
+        text_align alignment, float max_width, bool *is_selected, bool *is_focused, bool *is_hovered )
+{
+    ImGui::PushID( text.c_str() );
+    ImGuiID itemId = GImGui->CurrentWindow->IDStack.back();
+    const auto color_segments = split_by_color( text );
+    std::stack<nc_color> color_stack;
+    color_stack.push( color );
+    size_t chars_per_line = size_t( max_width );
+    if( chars_per_line == 0 ) {
+        chars_per_line = SIZE_MAX;
+    }
+    float cursor_start_x = ImGui::GetCursorPosX();
+#if defined(WIN32) || defined(TILES)
+    size_t char_width = size_t( ImGui::CalcTextSize( " " ).x );
+    chars_per_line /= char_width;
+#endif
+    if( alignment != text_align::Left ) {
+        std::string str_raw = remove_color_tags( text );
+        int fullWidth = ImGui::GetContentRegionAvail().x;
+        float textWidth = ImGui::CalcTextSize( str_raw.c_str() ).x;
+        if( alignment == text_align::Right ) {
+            ImGui::SetCursorPosX( ImGui::GetCursorPosX() + fullWidth - textWidth - 2 );
+        } else if( alignment == text_align::Center ) {
+            ImGui::SetCursorPosX( ImGui::GetCursorPosX() + ( float( fullWidth ) / 2 ) - ( textWidth / 2 ) );
+        }
+    }
+    if( is_selected != nullptr ) {
+        ImGui::Selectable( "", is_selected );
+        ImGui::SameLine( 0, 0 );
+    }
+
+    int i = 0;
+    size_t current_x = 0;
+    for( auto seg : color_segments ) {
+        if( seg.empty() ) {
+            continue;
+        }
+
+        if( seg[0] == '<' ) {
+            const color_tag_parse_result::tag_type type =
+                update_color_stack( color_stack, seg, report_color_error::yes );
+            if( type != color_tag_parse_result::non_color_tag ) {
+                seg = rm_prefix( seg );
+            }
+        }
+
+        color = color_stack.empty() ? color : color_stack.top();
+        for( size_t current_seg_index = 0; current_seg_index < seg.length(); ) {
+
+            if( i++ > 0 ) {
+                if( current_x != 0 ) {
+                    ImGui::SameLine( 0, 0 );
+                } else if( alignment == text_align::Left ) {
+                    ImGui::SetCursorPosX( cursor_start_x );
+                }
+            }
+            size_t chars_to_print = seg.length() - current_seg_index;
```

Comment:
`std::string::length` and `std::string::substr` won't work for unicode strings. Please use `utf8_wrapper::display_width`  and `utf8_wrapper::substr_display` from `catacharset.h` instead.

Reply:
probably going to rewrite the wrapping anyway, i discovered foldstring after i wrote it, so I think this all will go away.

Reply:
OK, wrapping is re-written, this should be resolved